src/re/correlated_field.py

```diff
@@ -27,6 +29,96 @@ def hartley(p, axes=None):
+    from jaxbind.contrib.jaxducc0 import get_healpix_sht
+
+    jsht = get_healpix_sht(nside, lmax, mmax, 0, nthreads)
+
+    def f(inp):
+        # Explicitly move axes around, may introduce unnecessary copies
+        inp = jnp.moveaxis(inp, axis, -1)
+        shp = inp.shape
+        inp = inp.reshape((-1, shp[-1]))
+
+        def trafo(x):
+            return np.sqrt(4 * np.pi) * jsht(x[jnp.newaxis])[0][0]
+
+        res = vmap(trafo)(inp).reshape(shp[:-1] + (-1,))
+        return jnp.moveaxis(res, -1, axis)
+
+    return f
+
+
+def _unique_mode_distributor(m_length):
+    # Construct an array of unique mode lengths
+    uniqueness_rtol = 1e-12
+    um = jnp.unique(m_length)
+    tol = uniqueness_rtol * um[-1]
+    um = um[jnp.diff(jnp.append(um, 2 * um[-1])) > tol]
+    # Group modes based on their length and store the result as power
+    # distributor
+    binbounds = 0.5 * (um[:-1] + um[1:])
+    m_length_idx = jnp.searchsorted(binbounds, m_length)
+    m_count = jnp.bincount(m_length_idx.ravel(), minlength=um.size)
+    if jnp.any(m_count == 0) or um.shape != m_count.shape:
+        raise RuntimeError("invalid harmonic mode(s) encountered")
+    return m_length_idx, um, m_count
+
+
+def get_spherical_mode_distributor(nside: int, lmax=None, mmax=None):
+    """Get the unique lengths of Spherical harmonic modes, a mapping from a mode
+    to its length index and the multiplicity of each unique mode length.
+
+    Parameters
+    ----------
+    nside : int
+        Nside of the HEALPix sphere for which the associated harmonic space is
+        constructed
+    lmax : int, optional
+        The maximum :math:`l` value of any spherical harmonic coefficient
+        :math:`a_{lm}` that is represented by this object.
+        Must be :math:`\\ge 0`. If not supplied, it is set to `nside * 2`
+    mmax : int, optional
+        The maximum :math:`m` value of any spherical harmonic coefficient
+        :math:`a_{lm}` that is represented by this object.
+        If not supplied, it is set to `lmax`.
+        Must be :math:`\\ge 0` and :math:`\\le` `lmax`.
+
+    Returns
+    -------
+    mode_length_idx : jnp.ndarray
+        Index in power-space for every mode in harmonic-space. Can be used to
+        distribute power from a power-space to the full harmonic grid.
+    unique_mode_length : jnp.ndarray
+        Unique length of spherical modes.
+    mode_multiplicity : jnp.ndarray
+        Multiplicity for each unique mode length.
+    """
+    if lmax is None:
+        lmax = 2 * nside
+    lmax = int(lmax)
+    if lmax < 0:
+        raise ValueError("lmax must be >=0.")
+    if mmax is None:
+        mmax = lmax
+    mmax = int(mmax)
+    if mmax < 0 or mmax > lmax:
+        raise ValueError("mmax must be >=0 and <=lmax.")
+    size = (lmax + 1) ** 2 - (lmax - mmax) * (lmax - mmax + 1)
+
+    ldist = np.empty((size,), dtype=np.float64)
+    ldist[0 : lmax + 1] = np.arange(lmax + 1, dtype=np.float64)
```

Comment:
very nit-picky: could we make the `dtype` an optional argument and default to `np.float64`?